packages/client/src/links/dedupeLink.ts

```diff
@@ -21,31 +20,33 @@ export function dedupeLink<
       const key = JSON.stringify([op.path, op.input]);
       if (pending[key]) {
         // console.log('hooking into pending', { op });
-        return observable((observer) => pending[key].subscribe(observer));
+        return new Observable((observer) => pending[key].subscribe(observer));
       }
 
-      const shared$ = observable((observer) => {
-        function reset() {
-          delete pending[key];
-        }
-        const next$ = next(op).subscribe({
-          next(v) {
-            observer.next(v);
-          },
-          error(e) {
+      const shared$ = new Observable<OperationResult<TRouter, unknown>>(
+        (observer) => {
+          function reset() {
+            delete pending[key];
+          }
+          const next$ = next(op).subscribe({
+            next(v) {
+              observer.next(v);
+            },
+            error(e) {
+              reset();
+              observer.error(e);
+            },
+            complete() {
+              reset();
```

Comment:
note: (you don't need to change this here) but even with your previous observable, when `observer.error` or `observer.complete` is called, The teardown below is called automatically, so `reset()` is technically called twice. It doesn't seem like that's ever mattered, just calling it out.